greymatter-io/terraform-greymatter-ecs
# Add dashboard & catalog

This PR does a couple of things:

- adds dashboard and catalog.
- refactors the services with sidecars (now control-api, dashboard, and catalog) within the same task definition - this makes sense logically, is fewer resources, and allows the service to talk to its sidecar over localhost (which helps minimize the problem of the complex way we're mounting certs)
- adds ability to specify service version in variable `versions` which is a map:

```tf
variable "versions" {
  default = {
    gm-proxy = "1.5.0"
    gm-control = "1.5.1-dev"
    gm-control-api = "1.5.0"
    gm-dashboard = "4.0.0"
    gm-catalog = "1.0.7"
  }
  type = map(string)
}

```


For now, catalog doesn't support using REST to talk to control, so any cards on the dashboard will stay red (issue filed [here](https://github.com/greymatter-io/gm-catalog/issues/375)). But the resources are here and ready.


To test in bs:

Checkout branch `ecs_module_test`.  Go to the ecs module at tf/resources/ecs_module_test.tf and fill in the values for `docker_gm_credentials`, `aws_secret_access_key` and `aws_access_key_id`. Then cd into `tf/resources` and run:

```
terraform get
terraform apply
```

When everything is created, copy the edge load balancer dns name and paste it into the following block to configure the cli:

```bash
export GREYMATTER_API_HOST={edgs_dns}
export GREYMATTER_API_INSECURE=true
export GREYMATTER_API_PREFIX=/services/control-api/latest
export GREYMATTER_API_SSL=true
export GREYMATTER_API_SSLCERT=gm/certs/edge/quickstart.crt
export GREYMATTER_API_SSLKEY=gm/certs/edge/quickstart.key
export GREYMATTER_CONSOLE_LEVEL=debug
```

In a browser you should be able to visit `https://{edge-dns}/services/control-api/latest` and see the control-api docs. Using the cli run `./gm/mesh/apply.sh`. This will apply the rest of the mesh configs. Give it a second to discover and then try navigating to the following:

```bash
https://{edge-dns}/services/control-api/latest
https://{edge-dns}/services/catalog/latest
https://{edge-dns}
```

The last should be the dashboard. There is one preconfigured card to Control API, but it will stay red because of the catalog issue mentioned above.

To tear it down,

```
terraform destroy -target=module.greymatter
```

You'll have to manually delete the autoscaling group in order for everything to properly destroy.
